usr/src/uts/common/fs/zfs/metaslab.c

```diff
@@ -3400,13 +3462,425 @@ metaslab_check_free(spa_t *spa, const blkptr_t *bp)
 		uint64_t size = DVA_GET_ASIZE(&bp->blk_dva[i]);
 		metaslab_t *msp = vd->vdev_ms[offset >> vd->vdev_ms_shift];
 
-		if (msp->ms_loaded)
+		mutex_enter(&msp->ms_lock);
+		if (msp->ms_loaded) {
 			range_tree_verify(msp->ms_tree, offset, size);
+			if (msp->ms_trimming_ts) {
+				range_tree_verify(msp->ms_trimming_ts,
+				    offset, size);
+			}
+#ifdef	DEBUG
+			VERIFY3P(&msp->ms_lock, ==, msp->ms_cur_ts->rt_lock);
+			range_tree_verify(msp->ms_cur_ts, offset, size);
+			if (msp->ms_prev_ts != NULL) {
+				VERIFY3P(&msp->ms_lock, ==,
+				    msp->ms_prev_ts->rt_lock);
+				range_tree_verify(msp->ms_prev_ts, offset,
+				    size);
+			}
+#endif
+		}
 
 		range_tree_verify(msp->ms_freeingtree, offset, size);
 		range_tree_verify(msp->ms_freedtree, offset, size);
 		for (int j = 0; j < TXG_DEFER_SIZE; j++)
 			range_tree_verify(msp->ms_defertree[j], offset, size);
+		mutex_exit(&msp->ms_lock);
 	}
 	spa_config_exit(spa, SCL_VDEV, FTAG);
 }
+
+/*
+ * Trims all free space in the metaslab. Returns the root TRIM zio (that the
```

Comment:
I'll reword the comment to describe the function more accurately. This comment is kind of an amalgam of how it used to function (trim everything) the later tweak that introduced the limiting function to `zfs_max_bytes_per_trim`.